internal/xfn/cached/cached_runner.go

```diff
@@ -0,0 +1,387 @@
+	// If we don't have a cache key we can't cache the response. Just send
+	// it on. This should never happen.
+	if req.GetMeta().GetTag() == "" {
+		return r.wrapped.RunFunction(ctx, name, req)
+	}
+
+	key := filepath.Join(name, req.GetMeta().GetTag())
+	log := r.log.WithValues("name", name, "cache-key", key)
+
+	rsp, err := r.wrapped.RunFunction(ctx, name, req)
+	if err != nil {
+		return rsp, err
+	}
+
+	// Start timing the cache write _after_ we make the request.
+	start := time.Now()
+
+	ttl := rsp.GetMeta().GetTtl().AsDuration()
+	if ttl == 0 {
+		return rsp, nil
+	}
+
+	// Not all filesystems have btime, and Go doesn't expose a simple
+	// interface to get it. We just write our own at the beginning of the
+	// file. We use basic binary encoding (as opposed to gob or protobuf)
+	// because the fixed length makes it easy to slice the file into the
+	// timestamp header and the protobuf message body.
+	deadline := time.Now().Add(ttl)
+	header := make([]byte, 8)
+	binary.LittleEndian.PutUint64(header, uint64(time.Now().Add(ttl).Unix())) //nolint:gosec // False positive for G115.
+
+	msg, err := proto.Marshal(rsp)
+	if err != nil {
+		log.Info("RunFunctionResponse cache write error", "err", err)
+		r.metrics.Error(name)
+		return rsp, nil
+	}
+
+	if err := r.fs.MkdirAll(name, 0o700); err != nil {
+		log.Info("RunFunctionResponse cache write error", "err", err)
+		r.metrics.Error(name)
+		return rsp, nil
+	}
+
+	if err := r.fs.WriteFile(key, append(header, msg...), 0o600); err != nil { //nolint:makezero // We want deadline to be padded to 8 bytes.
+		log.Info("RunFunctionResponse cache write error", "err", err)
+		r.metrics.Error(name)
+		return rsp, nil
+	}
+
+	log.Debug("RunFunctionResponse cache write", "deadline", deadline, "bytes", len(header)+len(msg))
+	r.metrics.Write(name)
+	r.metrics.WriteDuration(name, time.Since(start))
+	r.metrics.WroteBytes(name, len(header)+len(msg))
+	return rsp, nil
+}
+
+// GarbageCollectFiles runs every interval until the supplied context is
+// cancelled. It garbage collects cached responses with expired deadlines.
+func (r *FileBackedRunner) GarbageCollectFiles(ctx context.Context, interval time.Duration) {
+	t := time.NewTicker(interval)
+	defer t.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			r.log.Debug("Stopping RunFunctionResponse cache garbage collector", "error", ctx.Err())
+			return
+		case <-t.C:
+			if _, err := r.GarbageCollectFilesNow(ctx); err != nil {
+				r.log.Info("Cannot garbage collect cached RunFunctionResponses", "error", err)
+			}
+		}
+	}
+}
+
+// GarbageCollectFilesNow immediately garbage collects any cached responses with
+// expired deadlines.
+func (r *FileBackedRunner) GarbageCollectFilesNow(ctx context.Context) (int, error) {
+	collected := 0
+	iofs := afero.NewIOFS(r.fs)
+	err := fs.WalkDir(iofs, "/", func(path string, d fs.DirEntry, err error) error {
```

Comment:
Any reason to use this rather than `afero.Walk(r.fs, "/", ...)`?

Reply:
My understanding is `WalkDir` is a little more performant. Afero doesn't seem to have its own `WalkDir` implementation - instead they have an example that uses this `io/fs` implementation.

See comment at https://pkg.go.dev/path/filepath#Walk